src/library/scala/collection/SeqLike.scala

```diff
@@ -504,15 +504,19 @@ trait SeqLike[+A, +Repr] extends Any with IterableLike[A, Repr] with GenSeqLike[
    *  @return  A new $coll which contains the first occurrence of every element of this $coll.
    */
   def distinct: Repr = {
-    val b = newBuilder
-    val seen = mutable.HashSet[A]()
-    for (x <- this) {
-      if (!seen(x)) {
-        b += x
-        seen += x
+    val isImmutable = this.isInstanceOf[immutable.Seq[_]]
+    if (isImmutable && lengthCompare(1) <= 0) repr
+    else {
+      val b = newBuilder
+      val seen = new mutable.HashSet[A]()
+      var it = this.iterator
+      var different = false
+      while (it.hasNext) {
+        val next = it.next
+        if (seen.add(next)) b += next else different = true
       }
+      if (different) b.result() else repr
```

Comment:
This needs to check for `isImmutable` (see my previous comment). You can't return `repr` if it is not immutable.